circuit/gas_constraints.go

```diff
@@ -0,0 +1,141 @@
+	gas GasConstraints,
+	hFunc MiMC,
+	lastRoots [types.NbRoots]Variable) (newStateRoot Variable, err error) {
+	notEmptyTx := 1
+	NewAccountRoot := lastRoots[0]
+	var (
+		NewAccountAssetsRoot = gas.AccountInfoBefore.AssetRoot
+		deltas               [types.NbGasAssets]AccountAssetDeltaConstraints
+	)
+
+	for i := 0; i < types.NbGasAssets; i++ {
+		deltas[i] = AccountAssetDeltaConstraints{
+			BalanceDelta:             gas.AssetAmounts[i],
+			LpDelta:                  types.ZeroInt,
+			OfferCanceledOrFinalized: types.ZeroInt,
+		}
+	}
+	AccountsInfoAfter := UpdateGasAccount(api, gas.AccountInfoBefore, deltas)
+
+	// verify account asset node hash
+	for i := 0; i < types.NbGasAssets; i++ {
+		api.AssertIsLessOrEqual(gas.AccountInfoBefore.AssetsInfo[i].AssetId, LastAccountAssetId)
+		assetMerkleHelper := AssetIdToMerkleHelper(api, gas.AccountInfoBefore.AssetsInfo[i].AssetId)
+		hFunc.Reset()
+		hFunc.Write(
+			gas.AccountInfoBefore.AssetsInfo[i].Balance,
+			gas.AccountInfoBefore.AssetsInfo[i].LpAmount,
+			gas.AccountInfoBefore.AssetsInfo[i].OfferCanceledOrFinalized,
+		)
+		assetNodeHash := hFunc.Sum()
+		// verify account asset merkle proof
+		hFunc.Reset()
+		types.VerifyMerkleProof(
+			api,
+			notEmptyTx,
+			hFunc,
+			NewAccountAssetsRoot,
+			assetNodeHash,
+			gas.MerkleProofsAccountAssetsBefore[i][:],
+			assetMerkleHelper,
+		)
+		hFunc.Reset()
+		hFunc.Write(
+			AccountsInfoAfter.AssetsInfo[i].Balance,
+			AccountsInfoAfter.AssetsInfo[i].LpAmount,
+			AccountsInfoAfter.AssetsInfo[i].OfferCanceledOrFinalized,
+		)
+		assetNodeHash = hFunc.Sum()
+		hFunc.Reset()
+		// update merkle proof
+		NewAccountAssetsRoot = types.UpdateMerkleProof(
+			api, hFunc, assetNodeHash, gas.MerkleProofsAccountAssetsBefore[i][:], assetMerkleHelper)
+	}
+	// verify account node hash
+	api.AssertIsEqual(gas.AccountInfoBefore.AccountIndex, 1)
+	accountIndexMerkleHelper := AccountIndexToMerkleHelper(api, gas.AccountInfoBefore.AccountIndex)
+	hFunc.Reset()
+	hFunc.Write(
+		gas.AccountInfoBefore.AccountNameHash,
+		gas.AccountInfoBefore.AccountPk.A.X,
+		gas.AccountInfoBefore.AccountPk.A.Y,
+		gas.AccountInfoBefore.Nonce,
+		gas.AccountInfoBefore.CollectionNonce,
+		gas.AccountInfoBefore.AssetRoot,
+	)
+	accountNodeHash := hFunc.Sum()
+	// verify account merkle proof
+	hFunc.Reset()
+	types.VerifyMerkleProof(
+		api,
+		notEmptyTx,
+		hFunc,
+		NewAccountRoot,
+		accountNodeHash,
+		gas.MerkleProofsAccountBefore[:],
+		accountIndexMerkleHelper,
+	)
+	hFunc.Reset()
+	hFunc.Write(
+		AccountsInfoAfter.AccountNameHash,
+		AccountsInfoAfter.AccountPk.A.X,
+		AccountsInfoAfter.AccountPk.A.Y,
+		AccountsInfoAfter.Nonce,
+		AccountsInfoAfter.CollectionNonce,
+		NewAccountAssetsRoot,
+	)
+	accountNodeHash = hFunc.Sum()
+	hFunc.Reset()
+	// update merkle proof
+	NewAccountRoot = types.UpdateMerkleProof(api, hFunc, accountNodeHash, gas.MerkleProofsAccountBefore[:], accountIndexMerkleHelper)
+
+	hFunc.Reset()
+	hFunc.Write(
+		NewAccountRoot,
+		lastRoots[0],
```

Comment:
This should be lastRoots[1], and next line should be lastRoots[2]?

Reply:
yes, typos